### PR TITLE
Node controller rebased

### PR DESCRIPTION
**Pull request by ozdanborne**
## Description

Add node controller.

## Todos
- [x] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Add a Node Controller, which tracks nodes that have been created with an orchRef, and cleans them up from Calico when they are deleted in Kubernetes.
```

### CODE REVIEW

File: pkg/controllers/node/node_controller.go [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+			calicoNodes, err := calicoClient.Nodes().List(api.NodeMetadata{})
+			if err != nil {
+				return nil, err
+			}
+
+			// Iterate through and store the k8s nodes in our cache.
+			m := make(map[string]interface{})
+			for _, calicoNode := range calicoNodes.Items {
+				// find its kubernetes orchRef
+				k8sNodeName := getK8sNodeName(calicoNode)
```
Comment (caseydavenport) [inline pkg/controllers/node/node_controller.go]: We only want to add the node to the map if it has a corresponding k8sNodeName.

So, I think we need to wrap this in an `if k8sNodeName != ""`

File: pkg/controllers/node/node_controller.go [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+				m[k8sNodeName] = calicoNode.Metadata.Name
```
Comment (caseydavenport) [inline pkg/controllers/node/node_controller.go]: I think that, strictly speaking, this is wrong - it should be returning a map of key / value that matches what is stored in the resource cache.

In this case, that is a map of `kubernetesNodeName` to `""`

File: pkg/controllers/node/node_controller.go [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+func (c *NodeController) populateNodeLookupCache() error {
+	nodes, err := c.calicoClient.Nodes().List(api.NodeMetadata{})
+	if err != nil {
+		return err
+	}
+
+	c.nodeLookupCache.Lock()
+	for _, node := range nodes.Items {
+		if k8sNodeName := getK8sNodeName(node); k8sNodeName != "" {
+			c.nodeLookupCache.nodes[k8sNodeName] = nodeData{node.Metadata.Name}
```
Comment (caseydavenport) [inline pkg/controllers/node/node_controller.go]: What's the purpose of the `nodeData` intermediate struct? Should probably just be a string, right?

File: pkg/controllers/node/node_controller.go [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+}
+
+// syncToCalico syncs the given update to the Calico datastore.
+func (c *NodeController) syncToCalico(key string) error {
+	// Check if it exists in the controller's cache.
+	_, exists := c.k8sResourceCache.Get(key)
+	if !exists {
+		// The object no longer exists - delete from the datastore.
+		clog := log.WithField("node", key)
+		c.nodeLookupCache.Lock()
```
Comment (caseydavenport) [inline pkg/controllers/node/node_controller.go]: Probably should use `RLock()` and `RUnlock()` for read operations

File: pkg/controllers/node/node_controller.go [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+	// Create a Node watcher.
+	listWatcher := corecache.NewListWatchFromClient(k8sClientset.Core().RESTClient(), "nodes", "", fields.Everything())
+
+	// Bind the Calico cache to kubernetes cache with the help of an informer. This way we make sure that
+	// whenever the kubernetes cache is updated, changes get reflected in the Calico cache as well.
+	_, informer := corecache.NewIndexerInformer(listWatcher, &v1.Node{}, 0, corecache.ResourceEventHandlerFuncs{
+		DeleteFunc: func(obj interface{}) {
+			nodeName, err := extractNodeName(obj)
+			if err != nil {
+				log.WithError(err).Errorf("Error while converting %#v to k8s node", nodeName)
```
Comment (caseydavenport) [inline pkg/controllers/node/node_controller.go]: AFAICT, whenever `extractNodeName` returns an error, the `nodeName` is an empty string, so this message shouldn't include it.

Also, it means we should return early here so we don't call `Delete("")`

File: pkg/controllers/node/node_controller.go [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+			}
+			log.Debugf("Got DELETE event for node: %s", nodeName)
+			k8sResourceCache.Delete(nodeName)
+		},
+
+		AddFunc: func(obj interface{}) {
+			log.Debug("Add func.")
+			nodeName, err := extractNodeName(obj)
```
Comment (caseydavenport) [inline pkg/controllers/node/node_controller.go]: Same here

File: pkg/controllers/node/node_controller.go (hunk 5 above)
Comment (caseydavenport) [reply pkg/controllers/node/node_controller.go]: We can still log out the `obj` we received.

File: pkg/controllers/node/node_controller.go (hunk 3 above)
Comment (ozdanborne) [reply pkg/controllers/node/node_controller.go]: Did I implement this on the wrong cache then? https://github.com/projectcalico/kube-controllers/pull/163#discussion_r143587839

File: tests/fv/fv_test.go [hunk 7]
```diff
@@ -81,6 +81,101 @@ var _ = Describe("PolicyController", func() {
 		apiserver.Stop()
 	})
 
+	Context("nodes", func() {
+		// It("should delete a node that exists at start of day", func() {})
```
Comment (caseydavenport) [inline tests/fv/fv_test.go]: Are these "TODOs"?

File: tests/fv/fv_test.go [hunk 8]
```diff
@@ -81,6 +81,101 @@ var _ = Describe("PolicyController", func() {
...
+				},
+			}
+			_, err = calicoClient.Nodes().Create(cn)
+			Expect(err).NotTo(HaveOccurred())
+
+			k8sClient.CoreV1().Nodes().Delete("k8snode", &metav1.DeleteOptions{})
+			Eventually(func() *api.Node {
+				node, _ := calicoClient.Nodes().Get(api.NodeMetadata{Name: "calnode"})
+				return node
+			}, time.Second*15, 500*time.Millisecond).Should(BeNil())
```
Comment (caseydavenport) [inline tests/fv/fv_test.go]: This duration is longer than the reconciler period, which means a reconciler delete might cause this test to pass.

There's not way to guarantee the reconciler isn't doing this, but we can at least make the test _flake_ if the reconciler is to blame.

Let's just drop this timeout down to one or two seconds - that should be plenty of time.

File: tests/fv/fv_test.go [hunk 9]
```diff
@@ -81,6 +81,101 @@ var _ = Describe("PolicyController", func() {
 		apiserver.Stop()
 	})
 
+	Context("nodes", func() {
+		// It("should delete a node that exists at start of day", func() {})
+		// It("should delete nodes that nodeRef unexisting node in k8s"), func() {})
+		// It("should clean up dead data for nodes"
+		It("should be removed in response to a k8s node delete", func() {
```
Comment (caseydavenport) [inline tests/fv/fv_test.go]: nit: the "should" clauses here are worded strangely. I think they are assuming "It" is the node object, when I think "It" should be the controller.

e.g. for this one,  "It should delete Calico nodes in response to deleting a k8s node"

File: tests/fv/fv_test.go [hunk 10]
```diff
@@ -81,6 +81,101 @@ var _ = Describe("PolicyController", func() {
...
+			}, time.Second*15, 500*time.Millisecond).Should(BeNil())
+		})
+
+		It("should not be removed if they don't have any orchrefs.", func() {
+			cn := &api.Node{
+				Metadata: api.NodeMetadata{
+					Name: "calnode",
+				},
+				Spec: api.NodeSpec{
+					OrchRefs: []api.OrchRef{},
```
Comment (caseydavenport) [inline tests/fv/fv_test.go]: What about an OrchRef from another orchestrator? That might be a more useful test than this one.

File: tests/fv/fv_test.go [hunk 11]
```diff
@@ -81,6 +81,101 @@ var _ = Describe("PolicyController", func() {
 		apiserver.Stop()
```
Comment (caseydavenport) [inline tests/fv/fv_test.go]: Are there any "resiliency" type tests we want to add for nodes?

File: tests/fv/fv_test.go (hunk 9 above)
Comment (ozdanborne) [reply tests/fv/fv_test.go]: I thought this was correct because these are all within `Context("nodes" )`

Comment (caseydavenport) [reply tests/fv/fv_test.go]: Yeah perhaps, I hadn't seen that.

However, I'd also suggest tweaking the context to "Node controller" to match the other contexts in this file

File: tests/fv/fv_test.go [hunk 12]
```diff
@@ -81,6 +87,237 @@ var _ = Describe("PolicyController", func() {
...
+				},
+			}
+			_, err = calicoClient.Nodes().Create(cn)
+			Expect(err).NotTo(HaveOccurred())
+
+			k8sClient.CoreV1().Nodes().Delete(kNodeName, &metav1.DeleteOptions{})
+			Eventually(func() *api.Node {
+				node, _ := calicoClient.Nodes().Get(api.NodeMetadata{Name: cNodeName})
+				return node
+			}, time.Second*15, 500*time.Millisecond).Should(BeNil())
```
Comment (caseydavenport) [inline tests/fv/fv_test.go]: Should drop this timeout down to 2 seconds as well

File: tests/fv/fv_test.go [hunk 13]
```diff
@@ -81,6 +87,237 @@ var _ = Describe("PolicyController", func() {
...
+				},
+			}
+			_, err := calicoClient.Nodes().Create(cn)
+			Expect(err).NotTo(HaveOccurred())
+
+			k8sClient.CoreV1().Nodes().Delete(kNodeName, &metav1.DeleteOptions{})
+			Eventually(func() *api.Node {
+				node, _ := calicoClient.Nodes().Get(api.NodeMetadata{Name: cNodeName})
+				return node
+			}, time.Second*2, 500*time.Millisecond).Should(BeNil())
```
Comment (caseydavenport) [inline tests/fv/fv_test.go]: This one should stay at 15s.

Any test that relies on the reconciler needs a longer timeout, any test that doesn't needs a short one.

**Comment by caseydavenport** [comment]
@ozdanborne once you get the tests passing let's squash this bad boy and get it merged.

**Pull request merged**